backend/api/public/attrs/crud.py

```diff
@@ -28,74 +33,75 @@ def add_str_attr(
         db.add(new_key)
 
     # Now, add the new EAV string attribute
-    pulse_str_attr = PulseStrAttrs(key=key, value=value, pulse_id=pulse_id)
+    pulse_str_attr = PulseAttrsStr(key=key, value=value, pulse_id=pulse_id)
     db.add(pulse_str_attr)
 
     db.commit()
     db.refresh(pulse)
-    return pulse
+    return PulseRead.from_orm(pulse)
 
 
 def read_pulse_attrs(
     pulse_id: int,
     db: Session = Depends(get_session),
-) -> list[dict[str, str]] | None:
+) -> list[PulseAttrsStrRead]:
     """Get all the keys for a pulse with id pulse_id."""
     pulse = db.get(Pulse, pulse_id)
     if not pulse:
-        return None
+        raise PulseNotFoundError(pulse_id=pulse_id)
 
-    statement = select([PulseStrAttrs.key, PulseStrAttrs.value]).where(
-        PulseStrAttrs.pulse_id == pulse_id,
+    statement = select(PulseAttrsStr).where(
+        PulseAttrsStr.pulse_id == pulse_id,
     )
-    results = db.execute(statement).fetchall()
-
-    return [{"key": row[0], "value": row[1]} for row in results]
+    attrs = db.exec(statement).all()
+    return [PulseAttrsStrRead.from_orm(attr) for attr in attrs]
 
 
 def read_all_keys(
     db: Session = Depends(get_session),
 ) -> list[str]:
     """Get all unique keys."""
-    return [key[0] for key in db.query(PulseKeyRegistry.key).all()]
+    statement = select(PulseKeyRegistry.key).distinct()
+    return db.exec(statement).all()
 
 
 def read_all_values_on_key(
     key: str,
     db: Session = Depends(get_session),
-) -> list[str]:
+) -> list[StrictStr]:
     """Get all unique values associated with a key."""
-    statement = select(PulseStrAttrs.value).where(PulseStrAttrs.key == key).distinct()
-    return db.execute(statement).scalars().all()
+    statement = select(PulseAttrsStr.value).where(PulseAttrsStr.key == key).distinct()
+    return db.exec(statement).all()
 
 
 def filter_on_key_value_pairs(
-    key_value_pairs: list[dict[str, str]],
+    kv_pairs: list[PulseAttrsStrRead],
     db: Session = Depends(get_session),
 ) -> list[int]:
     """Get all pulses that match the key-value pairs."""
     # Initialize a list to hold pulse_ids for each condition
-    pulse_ids_list = []
+    pulse_ids_list: list[set[int]] = []
 
     # If no filters applied, select all pulses
-    if len(key_value_pairs) == 0:
-        return db.execute(select(Pulse.pulse_id)).scalars().all()
+    if len(kv_pairs) == 0:
+        pulses = db.exec(select(Pulse.pulse_id)).all()
+        if pulses[0] is None:
+            return []
 
-    for kv in key_value_pairs:
-        key = kv.get("key")
-        value = kv.get("value")
+    for kv in kv_pairs:
+        key = kv.key
+        value = kv.value
 
         # Perform query for this key-value pair
-        statement = select(PulseStrAttrs.pulse_id).where(
-            and_(PulseStrAttrs.key == key, PulseStrAttrs.value == value),
+        statement = (
+            select(PulseAttrsStr.pulse_id)
+            .where(PulseAttrsStr.key == key)
```

Comment:
Why not use `kv` directly?
```suggestion
            .where(PulseAttrsStr.key == kv.key)
```